ta/pkcs11/src/pkcs11_token.c

```diff
@@ -80,3 +81,96 @@ void pkcs11_deinit(void)
 	for (id = 0; id < TOKEN_COUNT; id++)
 		close_persistent_db(get_token(id));
 }
+
+uint32_t entry_ck_slot_list(uint32_t ptypes, TEE_Param *params)
+{
+	const uint32_t exp_pt = TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INOUT,
+						TEE_PARAM_TYPE_NONE,
+						TEE_PARAM_TYPE_MEMREF_OUTPUT,
+						TEE_PARAM_TYPE_NONE);
+	TEE_Param *out = &params[2];
+	const size_t out_size = sizeof(uint32_t) * TOKEN_COUNT;
+	uint8_t *id = NULL;
+	uint32_t n = 0;
+
+	if (ptypes != exp_pt ||
+	    params[0].memref.size != TEE_PARAM0_SIZE_MIN)
+		return PKCS11_CKR_ARGUMENTS_BAD;
+
+	if (out->memref.size < out_size) {
+		out->memref.size = out_size;
+
+		if (out->memref.buffer)
+			return PKCS11_CKR_BUFFER_TOO_SMALL;
+		else
+			return PKCS11_CKR_OK;
```

Comment:
Why is `PKCS11_CKR_OK` returned here? I would expect `PKCS11_CKR_BUFFER_TOO_SMALL` regardless of the attached buffer.

Reply:
The spec expects `C_GetSlotList()` to return `CKR_OK` when output buffer reference is `NULL`, and `CKR_BUFFER_TOO_SMALL` only when output buffer reference is not `NULL` but buffer size is too small.

I reflect that in the TA. But note TA could return `PKCS11_CKR_BUFFER_TOO_SMALL` in both cases and let `libckteec` to set to appropriate return value.

Reply:
Yes, AFAICS there are two cases:
1. Buffer allocated by the library in `ckteec_alloc_shm()`
Will always return a valid allocation with set `out->memref.buffer` or fail early in the libary.
2. Buffer passed in via `ckteec_register_shm()`
Here we could handle this in the library and remove the handling from the TA.

Reply:
`libckteec` has been changed to allocated a buffer only upon caller of `C_GetSlotList()` providing a buffer (and non-null ID count value), see https://github.com/OP-TEE/optee_client/pull/190/files#diff-4e74e61ecbbd33455d2492ac57005e97R73-R78.

Reply:
I checked the code against the spec and saw that you pointed out an issue. Not a severe one but the current implementation does not strictly follow the spec on the topic of this thread (`C_GetSlotList()` and null/non-null buffer/count values).

Fixing that, in my opion, can be done either in both TA and libckteec, or only in libckteec, hence created https://github.com/OP-TEE/optee_client/pull/191. @Emantor, please have a look.